File: school-finder-scraper/scrapers/men_scraper.py
from scrapers.base_scraper import BaseScraper
import json

class MENScraper(BaseScraper):
    """
    Scraper do pobierania danych ze strony Ministerstwa Edukacji Narodowej (MEN)
    """
    def __init__(self, config):
        super().__init__(config)
        self.base_url = config.get('men_base_url', 'https://rspo.gov.pl')
        self.search_url = config.get('men_search_url', 'https://rspo.gov.pl/szkoly-i-placowki/wyszukiwarka')
    
    def get_school_list(self, region=None, school_type=None, limit=None):
        """
        Pobieranie listy szkół z wyszukiwarki MEN
        
        Args:
            region (str, optional): Region do filtrowania szkół
            school_type (str, optional): Typ szkoły do filtrowania
            limit (int, optional): Limit liczby szkół do pobrania
            
        Returns:
            list: Lista szkół w formacie [{name, rspo, url, ...}, ...]
        """
        # Parametry wyszukiwania
        params = {}
        if region:
            params['wojewodztwo'] = region
        if school_type:
            params['typ'] = school_type
        
        # Pobieranie strony z wynikami wyszukiwania
        response = self.fetch(self.search_url, params=params)
        
        # Parsowanie wyników
        # W rzeczywistej implementacji należy użyć XPathParser
        # Dla uproszczenia zwracamy przykładowe dane
        schools = [
            {
                'name': 'Szkoła Podstawowa nr 1',
                'rspo': '12345',
                'type': 'primary',
                'address': 'ul. Szkolna 1, 00-001 Warszawa',
                'url': f"{self.base_url}/szkola/12345"
            },
            {
                'name': 'Liceum Ogólnokształcące nr 1',
                'rspo': '67890',
                'type': 'high_school',
                'address': 'ul. Licealna 1, 00-002 Warszawa',
                'url': f"{self.base_url}/szkola/67890"
            }
        ]
        
        # Zastosowanie limitu, jeśli podano
        if limit and isinstance(limit, int):
            schools = schools[:limit]
        
        return schools
    
    def get_school_details(self, rspo):
        """
        Pobieranie szczegółowych informacji o szkole na podstawie numeru RSPO
        
        Args:
            rspo (str): Numer RSPO szkoły
            
        Returns:
            dict: Szczegółowe informacje o szkole
        """
        # URL do strony ze szczegółami szkoły
        url = f"{self.base_url}/szkola/{rspo}"
        
        # Pobieranie strony ze szczegółami
        response = self.fetch(url)
        
        # Parsowanie szczegółów
        # W rzeczywistej implementacji należy użyć XPathParser
        # Dla uproszczenia zwracamy przykładowe dane
        school_details = {
            'rspo': rspo,
            'name': 'Szkoła Przykładowa',
            'type': 'primary',
            'address': 'ul. Przykładowa 1, 00-001 Warszawa',
            'description': 'Opis szkoły przykładowej',
            'principal': 'Jan Kowalski',
            'contact': {
                'phone': '+48123456789',
                'email': 'szkola@example.com',
                'website': 'https://szkola-przykladowa.pl'
            },
            'location': {
                'lat': 52.2297,
                'lng': 21.0122
            },
            'region': 'mazowieckie',
            'county': 'Warszawa',
            'community': 'Warszawa'
        }
        
        return school_details

File: src/lib/scraper/scrapers/men-scraper.ts
/**
 * Scraper dla danych z MEN (Ministerstwo Edukacji Narodowej)
 */

import { BaseScraper } from './base-scraper';
import { ScraperConfig } from '../index';
import { Logger, setupLogger } from '../utils/logger';

interface School {
  name: string;
  rspo: string;
  type: string;
  address: string;
  url?: string;
}

export class MENScraper extends BaseScraper {
  private baseUrl: string;
  private searchUrl: string;
  private logger: Logger;
  
  constructor(config: ScraperConfig) {
    super(config);
    this.baseUrl = config.menBaseUrl || 'https://rspo.gov.pl';
    this.searchUrl = config.menSearchUrl || 'https://rspo.gov.pl/szkoly-i-placowki/wyszukiwarka';
    this.logger = setupLogger(config.logLevel);
  }
  
  /**
   * Pobieranie listy szkół z MEN
   */
  async getSchoolList(region?: string, schoolType?: string, limit?: number): Promise<School[]> {
    this.logger.info(`Fetching school list from MEN${region ? ` for region ${region}` : ''}`);
    
    const params: Record<string, any> = {};
    if (region && region !== 'all') {
      params.wojewodztwo = region;
    }
    if (schoolType && schoolType !== 'all') {
      params.typ = schoolType;
    }
    
    try {
      const response = await this.fetch(this.searchUrl, 'GET', params);
      
      // W rzeczywistej implementacji, parsowalibyśmy tutaj HTML
      // Dla uproszczenia, zwracamy przykładowe dane
      const schools: School[] = this.parseSchoolList(response.data);
      
      this.logger.info(`Found ${schools.length} schools on MEN website`);
      
      if (limit && limit > 0) {
        return schools.slice(0, limit);
      }
      
      return schools;
    } catch (error) {
      this.logger.error(`Failed to fetch school list from MEN: ${error instanceof Error ? error.message : String(error)}`);
      throw error;
    }
  }
  
  /**
   * Pobieranie szczegółów szkoły z MEN
   */
  async getSchoolDetails(rspo: string): Promise<any> {
    this.logger.info(`Fetching school details for RSPO: ${rspo}`);
    
    try {
      const response = await this.fetch(`${this.baseUrl}/szkola/${rspo}`);
      
      // W rzeczywistej implementacji, parsowalibyśmy tutaj HTML
      // Dla uproszczenia, zwracamy przykładowe dane
      const schoolDetails = this.parseSchoolDetails(response.data, rspo);
      
      return schoolDetails;
    } catch (error) {
      this.logger.error(`Failed to fetch school details for RSPO ${rspo}: ${error instanceof Error ? error.message : String(error)}`);
      throw error;
    }
  }
  
  /**
   * Parsowanie listy szkół z HTML
   */
  private parseSchoolList(html: string): School[] {
    // W rzeczywistej implementacji, parsowalibyśmy tutaj HTML
    // Dla uproszczenia, zwracamy przykładowe dane
    return [
      {
        name: 'Szkoła Podstawowa nr 1',
        rspo: '12345',
        type: 'primary',
        address: 'ul. Szkolna 1, 00-001 Warszawa'
      },
      {
        name: 'Liceum Ogólnokształcące nr 1',
        rspo: '67890',
        type: 'secondary',
        address: 'ul. Licealna 1, 00-002 Warszawa'
      }
    ];
  }
  
  /**
   * Parsowanie szczegółów szkoły z HTML
   */
  private parseSchoolDetails(html: string, rspo: string): any {
    // W rzeczywistej implementacji, parsowalibyśmy tutaj HTML
    // Dla uproszczenia, zwracamy przykładowe dane
    return {
      name: 'Szkoła Podstawowa nr 1',
      rspo: rspo,
      type: 'primary',
      address: 'ul. Szkolna 1, 00-001 Warszawa',
      description: 'Szkoła podstawowa z oddziałami integracyjnymi',
      principal: 'Jan Kowalski',
      contact: {
        phone: '123-456-789',
        email: 'sp1@example.com',
        website: 'https://sp1.example.com'
      },
      location: {
        lat: 52.2297,
        lng: 21.0122
      }
    };
  }
}